habitat-lab/habitat/tasks/rearrange/actions/actions.py

```diff
@@ -470,6 +470,218 @@ def step(self, *args, is_last_action, **kwargs):
+            # Sanitize the height
+            pos[1] = 0.0
+            cur_pos[i][1] = 0.0
+            goal_pos[i][1] = 0.0
+            end_pos.append(pos)
+
+        # Planar move distance clamped by NavMesh
+        move = []
+        for i in range(num_check_cylinder):
+            move.append((end_pos[i] - goal_pos[i]).length())
+
+        # Wrap the move direction if we use sliding
+        # Find the largest diff moving direction, which means that there is a collision in that cylinder
+        if compute_sliding:
+            max_idx = np.argmax(move)
+            move_vec = end_pos[max_idx] - cur_pos[max_idx]
+            new_end_pos = trans.translation + move_vec
+            new_target_trans = mn.Matrix4.from_(
+                target_rigid_state.rotation.to_matrix(), new_end_pos
+            )
+        else:
+            new_target_trans = None
+
+        # For detection of linear or angualr velocities
+        # There is a collision if the difference between the clamped NavMesh position and target position is too great for any point.
+        diff = len([v for v in move if v > self._collision_threshold])
+
+        if diff > 0:
+            return True, new_target_trans
+        else:
+            return False, new_target_trans
+
+    def update_base(self, if_rotation):
+        """
+        Update the base of the robot
+        if_rotation: if the robot is rotating or not
+        """
+        # Get the control frequency
+        ctrl_freq = self._sim.ctrl_freq
+        # Get the current transformation
+        trans = self.cur_articulated_agent.sim_obj.transformation
+        # Get the current rigid state
+        rigid_state = habitat_sim.RigidState(
+            mn.Quaternion.from_matrix(trans.rotation()), trans.translation
+        )
+        # Integrate to get target rigid state
+        target_rigid_state = self.base_vel_ctrl.integrate_transform(
+            1 / ctrl_freq, rigid_state
+        )
+        # Get the traget transformation based on the target rigid state
+        target_trans = mn.Matrix4.from_(
+            target_rigid_state.rotation.to_matrix(),
+            target_rigid_state.translation,
+        )
+        # We do sliding only if we allow the robot to do sliding and current
+        # robot is not rotating
+        compute_sliding = self._allow_dyn_slide and not if_rotation
+        # Check if there is a collision
+        did_coll, new_target_trans = self.collision_check(
+            trans, target_trans, target_rigid_state, compute_sliding
+        )
+        # Update the base
+        if new_target_trans is not None:
+            self.cur_articulated_agent.sim_obj.transformation = (
+                new_target_trans
+            )
+        else:
+            if did_coll:
+                self.cur_articulated_agent.sim_obj.transformation = trans
+            else:
+                self.cur_articulated_agent.sim_obj.transformation = (
+                    target_trans
+                )
```

Comment:
If `new_target_trans` returned by `collision_check` was always the correct final transform (see above suggestion), this could be reduced to:
```suggestion
    self.cur_articulated_agent.sim_obj.transformation = new_target_trans
```